[FEAT] 섹션 업데이트 API 구현

JSeHoone opened the pull request:
## ❗️ 이슈 번호
#26 

## 📝 작업 내용
**1. Section**
- Section의 정보를 업데이트 하기 위한 Setter 메서드 추가.
- title, orderNo 필드만 Setter 메서드로 제공.

**2. SectionController** 
- PathVariable로 SectionId를 받도록 구현하였습니다.
- 업데이트하고자 하는 title, orderNo를 Body값으로 받도록 구현하였습니다.
- HTTP post 메서드로 정의하였습니다.
- try catch 구문으로 존재하지 않는 Section을 업데이트 하는 경우 400에러를 발생하도록 구현하였습니다.

**3. SectionService, SectionServiceImpl**
- SectionService 인터페이스에 modifySection 메서드를 정의하였습니다.
- SectionServiceImpl에서 비즈니스 로직을 구현하였습니다.
- `@Transactional`을 이용하여 Section 객체의 변경이 일어나면 자동으로 업데이트 되도록 구현하였습니다.

**4. SectionUpdateRequestDTO, SectionUpdateResponseDTO**
- `@Validation`을 이용하여 요청에 필요한 필드들을 검증하였습니다.
- ResponseDTO에서는`@Getter`를 이용하여 직렬화되지 않는 문제를 해결하였습니다.

## 💭 주의 사항
1. 현재 title이랑 orderNo만 변경하는 로직이 구현되어있습니다. 
   1번 orderNo를 2번으로 변경하면 2번 orderNo를 가지고 있는 Section은 3번으로 넘어가는 로직은 아직 미구현 상태입니다. 

## 💡 리뷰 포인트
1. Service 부분에서 자동으로 업데이트 되도록 구현하였는데, 문제가 없는지 확인 부탁드립니다 !
2. 현재 OrderNo를 변경하는 경우 변경될 OrderNo를 +1해주도록 구현하면 해당 기능은 구현이 되는 건지 궁금합니다

JSeHoone committed:
feat(Section) : Update를 위한 Setter메서드 추가

JSeHoone committed:
feat : SectionController 내 updateSection 메서드 추가

JSeHoone committed:
feat : SectionService 내 ModifySection 메서드 추가

- `@Transactional`어노테이션으로 Section 변경을 탐지하여 자동으로 업데이트 되도록 만들었습니다.
- Section의 ID값을 기준으로 찾고 없으면 오류를 내뱉도록 합니다.

JSeHoone committed:
feat : Update Request, Response DTO 추가

- SectionUpdateRequestDTO : `@Validation` 라이브러리를 활용하여 요청 Body를 검증하였습니다.
- SectionUpdateResponseDTO : `@Validation` 라이브러리를 활용하여 응답 Body를 생성할 때 검증을 진행하였습니다.`Getter`어노테이션을 사용하여 직렬화 문제를 해결했습니다.

jjeongdong commented on src/main/java/com/example/projectlxp/section/controller/dto/response/SectionUpdateResponseDTO.java:
```diff
@@ -0,0 +1,25 @@
+public class SectionUpdateResponseDTO {
+
+    @NotNull(message = "Section ID 값은 필수 입니다.")
+    public Long sectionId;
+
+    @NotBlank(message = "섹션의 제목을 입력해주세요.")
+    public String title;
+
+    @NotNull(message = "Order No 값은 필수 입니다.")
+    public int orderNo;
+
+    public SectionUpdateResponseDTO(Long sectionId, String title, int orderNo) {
+        this.sectionId = sectionId;
+        this.title = title;
+        this.orderNo = orderNo;
+    }
+}
```
response dto인데 validation 검증이 필요한 이유가 궁금합니다

JSeHoone replied:
ResponseDTO를 생성하는 과정에서 누락될 수 있는 상황을 방지하려고 검증 어노테이션을 추가했었습니다 !

불필요하면 삭제하도록 하겠습니다 !

JSeHoone replied:
Response에서는 Validation이 적용되지 않으니 빼도록 !

jjeongdong commented on src/main/java/com/example/projectlxp/section/service/impl/SectionServiceImpl.java:
```diff
@@ -57,4 +58,28 @@ public SectionCreateResponseDTO registerSection(Long courseId, String title, int
+
+        // update Section
+        findSection.setTitle(title);
+        findSection.setOrderNo(orderNo);
```
`updateSection`이란 메서드를 두고 그 메서드안에서 set 메서드를 호출하면 더 가독성이 좋아질 것 같습니다.

jjeongdong commented on src/main/java/com/example/projectlxp/section/service/impl/SectionServiceImpl.java:
```diff
@@ -57,4 +58,28 @@ public SectionCreateResponseDTO registerSection(Long courseId, String title, int
+                new SectionUpdateResponseDTO(
+                        findSection.getId(), findSection.getTitle(), findSection.getOrderNo());
+
+        return response;
```
그냥 바로 return문으로 반환해도 좋을 것 같습니다.

jjeongdong commented on src/main/java/com/example/projectlxp/section/controller/SectionController.java:
```diff
@@ -40,4 +44,19 @@ public BaseResponse<SectionCreateResponseDTO> createSection(
+
+        } catch (IllegalArgumentException e) {
+            return BaseResponse.error(HttpStatus.BAD_REQUEST, "존재하지 않는 Section입니다.");
+        }
```
controller 레이어에 try-catch문을 두는 것 보다는 전역예외 처리기를 두고 거기서 모든 예외를 핸들링하는 것이 좋을 것 같습니다.

jjeongdong commented on src/main/java/com/example/projectlxp/section/service/SectionService.java:
```diff
@@ -13,4 +14,14 @@ public interface SectionService {
+     * @param orderNo 섹션의 순서
+     * @return SectionUpdateResponseDTO
+     */
+    public SectionUpdateResponseDTO modifySection(Long sectionId, String title, int orderNo);
```
불필요한 `public` 접근 제어자 삭제하면 코드가 더 깔끔해질 것 같습니다.

JSeHoone committed:
feat : updateSection 메서드 추가

- 기존 Service에서 setter로 필드 값을 변경해주었는데, 메서드로 변경하도록 기능 추가

JSeHoone committed:
refactor : try-catch 구문 삭제

- 추후에 전역으로 예외 처리를 할 예정이므로, try-catch 구문 삭제

JSeHoone committed:
refactor : 불필요한 public 접근제어자 삭제

JSeHoone committed:
refactor : updateSection 로직 리팩토링

- SectionEntity에 updateSection 메서드를 추가하여 코드 리팩토링
- SectionUpdateResponseDTO를 생성하자마자 바로 반환하도록 리팩토링

JSeHoone committed:
refactor : ResponseDTO의 Validation 라이브러리 제거

- ResponseDTO를 생성할 때는 검증을 따로 해주지 않기에 불필요한 코드를 제거하였습니다.

jjeongdong commented on src/main/java/com/example/projectlxp/section/service/impl/SectionServiceImpl.java:
```diff
@@ -57,4 +58,22 @@ public SectionCreateResponseDTO registerSection(Long courseId, String title, int
+        findSection.updateSection(title, orderNo);
+
+        // convert To SectionUpdateResponseDTO & return
+        return new SectionUpdateResponseDTO(findSection.getId(), " ", findSection.getOrderNo());
```
여기 부분 수정해야될 것 같습니다!

JSeHoone committed:
fix : 테스트로 넣은 " " 코드 수정
